paddle/framework/executor.h

```diff
@@ -0,0 +1,37 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "paddle/framework/framework.pb.h"
+#include "paddle/framework/op_info.h"
+#include "paddle/framework/scope.h"
+#include "paddle/framework/tensor.h"
+#include "paddle/platform/device_context_manager.h"
+
+namespace paddle {
+namespace framework {
+
+class Executor {
+ public:
+  explicit Executor(const std::vector<platform::Place>& places);
+  ~Executor() {}
+  void Run(const ProgramDesc&, Scope*, std::vector<Tensor>*);
```

Comment:
I don't remember how should users specify the variables to be returned?

Also, it seems that we need a `std::vector<Variable*>` here, because not all variables have value type `Tensor`?

Reply:
Indeed, Executor Run interface will take both inputs and outputs. In every mini-batch, new batch of data will feed into Executor, and some outputs which users interested in will be fetched out.  What users want to fetch should be Tensors.
So, the interface of Run may be like this:

```c++
typedef std::vector<std::pair<string, Tensor>> TensorList;
void Run(const ProgramDesc&, const TensorList& inputs, Scope*, TensorList* outputs);
```

Reply:
Agree with Yi, maybe return std::vector<Variable*> would be more appropriate.

I think `inputs` should be in `ProgramDesc` in the form of a fetch OP (i.e., not an argument of `Run`).

Reply:
Agree with should return `std::vector<Variable*>`.  
My point is maybe the targets still can be part of `Run`'s argument. @helinwang 
Here are some hints how tensorflow does.
> The tensor store remembers the tensors we choose to keep for the current run call. It is available to every op kernel.
```c++
class TensorStore {
 public:
  struct TensorAndKey {
    Tensor tensor;
    int64 id;
    string device_name;
}
 private:
  mutex lock_;

  // The tensors that will be saved to session state when this run completes.
  // A map from tensor string name to tensor.
  std::unordered_map<string, TensorAndKey> tensors_ GUARDED_BY(lock_);
};
```
(SessionState)[https://github.com/tensorflow/tensorflow/blob/ff2da4156bcde26a253fd98198c6f9389df921d9/tensorflow/core/framework/session_state.h#L58]

which is used by the [executor](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.h#L83).

Reply:
My reasoning for why input and output should be part of the `ProgramDesc` is because I think a program should contain input and output.